internal/controller/management_controller.go

```diff
@@ -92,101 +83,143 @@ func (r *ManagementReconciler) Update(ctx context.Context, management *hmc.Manag
+		if _, _, err := helm.ReconcileHelmRelease(ctx, r.Client, component.helmReleaseName, r.SystemNamespace, helm.ReconcileHelmReleaseOpts{
 			Values:          component.Config,
 			ChartRef:        template.Status.ChartRef,
 			DependsOn:       component.dependsOn,
 			TargetNamespace: component.targetNamespace,
 			CreateNamespace: component.createNamespace,
-		})
-		if err != nil {
-			errMsg := fmt.Sprintf("error reconciling HelmRelease %s/%s: %s", r.SystemNamespace, component.Template, err)
-			updateComponentsStatus(detectedComponents, &detectedProviders, detectedContracts, component.helmReleaseName, component.Template, template.Status.Providers, template.Status.CAPIContracts, errMsg)
+		}); err != nil {
+			errMsg := fmt.Sprintf("Failed to reconcile HelmRelease %s/%s: %s", r.SystemNamespace, component.helmReleaseName, err)
+			updateComponentsStatus(statusAccumulator, component, nil, errMsg)
 			errs = errors.Join(errs, errors.New(errMsg))
+
 			continue
 		}
 
 		if component.Template != hmc.CoreHMCName {
 			if err := r.checkProviderStatus(ctx, component.Template); err != nil {
-				updateComponentsStatus(detectedComponents, &detectedProviders, detectedContracts, component.helmReleaseName, component.Template, template.Status.Providers, template.Status.CAPIContracts, err.Error())
+				updateComponentsStatus(statusAccumulator, component, nil, fmt.Sprintf("Failed to check provider status: %s", err))
 				errs = errors.Join(errs, err)
 				continue
 			}
 		}
 
-		updateComponentsStatus(detectedComponents, &detectedProviders, detectedContracts, component.helmReleaseName, component.Template, template.Status.Providers, template.Status.CAPIContracts, "")
+		updateComponentsStatus(statusAccumulator, component, template, "")
 	}
 
+	management.Status.AvailableProviders = statusAccumulator.providers
+	management.Status.CAPIContracts = statusAccumulator.compatibilityContracts
+	management.Status.Components = statusAccumulator.components
 	management.Status.ObservedGeneration = management.Generation
-	management.Status.AvailableProviders = detectedProviders
-	management.Status.CAPIContracts = detectedContracts
-	management.Status.Components = detectedComponents
 	management.Status.Release = management.Spec.Release
+
 	if err := r.Status().Update(ctx, management); err != nil {
-		errs = errors.Join(errs, fmt.Errorf("failed to update status for Management %s: %w",
-			management.Name, err))
+		errs = errors.Join(errs, fmt.Errorf("failed to update status for Management %s: %w", management.Name, err))
 	}
+
 	if errs != nil {
 		l.Error(errs, "Multiple errors during Management reconciliation")
 		return ctrl.Result{}, errs
 	}
+
 	return ctrl.Result{}, nil
 }
 
+func (r *ManagementReconciler) cleanupRemovedComponents(ctx context.Context, management *hmc.Management) error {
+	var (
+		errs error
+		l    = ctrl.LoggerFrom(ctx)
+	)
+
+	for compName, compSt := range management.Status.Components {
```

Comment:
the management object status cannot be relied upon as a source of information in the current implementation. I suggest determining the current state based on HelmRelease objects instead